Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java

```diff
@@ -139,6 +140,9 @@ public DistanceType getDistanceType() {
     @Config(description = "The number of threads to use for training.")
     private int numThreads = 1;
 
+    @Config(description = "The nearest neighbour implementation factory to use.")
+    private NeighboursQueryFactory neighboursQueryFactory;
```

Comment:
I have set a default here, and have also set a default for `distType` to ensure these stay in sync. `distType` is also new in 4.3. Deprecating numThreads makes sense - the neighbour queries are the only part of Hdbscan that can be multithreaded. `distType` on the other hand, is referenced in other places in this class, so it doesn't feel right to deprecate that and access it through the query factory. I'm open to further discussion on this though.

Reply:
Don't set a default for `distType`, old config files will set the other field and fail validation in `postConfig`, it's fine for both `distance` and `distType` to not have defaults because of the upgrade path.

I'm not sure if we want to push it in either, but if it's in two places then it should validate that they are both the same in `postConfig`.

Reply:
OK, I think we're getting closer on this. Now, there are no defaults being set for these fields since they are handled in `postConfig`. There is also some validation incase there are differing `DistanceType` constants being used.